c_src/EiUtil.cc

```diff
@@ -0,0 +1,1689 @@
+           return formatTuple(buf, index);
+       
+       if(isBinary(type))
+           return formatBinary(buf, index);
+       
+       if(isString(type))
+           return formatString(buf, index);
+       
+       if(isList(type))
+           return formatList(buf, index);
+
+       if(isBig(buf, index))
+           return formatBig(buf, index);
+       
+       if(isNil(type))
+           return "[]";
+
+       std::stringstream os;
+       os << "?" << type << "size=" << getSizeInElements(buf, index);
+
+       return os.str();
+    )
+
+FN_DEF(std::string, formatAtom,
+       
+       std::ostringstream os;
+       os << getAtom(buf, index);
+       return os.str();
+    )
+
+FN_DEF(std::string, formatInteger,
+       
+       std::ostringstream os;
+       os << getLong(buf, index);
+       return os.str();
+    )
+
+FN_DEF(std::string, formatBig,
+       
+       std::ostringstream os;
+       EiUtil::Big big = getBig(buf, index);
+       os << (big.isSigned_ ? "-" : "") << big.val_;
+       return os.str();
+    )
+
+FN_DEF(std::string, formatFloat,
+
+       std::ostringstream os;
+       os << getFloat(buf, index);
+       return os.str();
+    )
+
+FN_DEF(std::string, formatTuple,
+
+       int arity = getTupleHeader(buf, index);
+       std::ostringstream os;
+
+       os << "{";
+       for(int iCell=0; iCell < arity; iCell++) {
+           os << formatTerm(buf, index);
+           if(iCell < arity-1)
+               os << ", ";
+       }
+       
+       os << "}";
+       
+       return os.str();
+    )
+
+FN_DEF(void, skipTuple,
+
+       int arity = getTupleHeader(buf, index);
+
+       for(int iCell=0; iCell < arity; iCell++)
+           skipLastReadObject(buf, index);
+    )
+
+FN_DEF(std::string, formatList,
+
+       int arity = getListHeader(buf, index);
+       std::ostringstream os;
+
+       os << "[";
+       for(int iCell=0; iCell < arity; iCell++) {
+           os << formatTerm(buf, index);
+           if(iCell < arity-1)
+               os << ", ";
+       }
+       
+       os << "]";
+       
+       return os.str();
+    )
+
+FN_DEF(void, skipList,
+
+       int arity = getListHeader(buf, index);
+
+       for(int iCell=0; iCell < arity; iCell++)
+           skipLastReadObject(buf, index);
+    )
+
+FN_DEF(std::string, formatBinary,
+
+       std::ostringstream os;
+       os << "<<";
+       std::vector<unsigned char> bin = getBinary(buf, index);
+       unsigned int size = bin.size();
+       for(unsigned int i=0; i < size; i++) {
+           os << (int)bin[i];
+           if(i < size-1)
+               os << ", ";
+       }
+       os << ">>";
+       
+       return os.str();
+    )
+
+FN_DEF(std::string, formatString,
+
+       std::ostringstream os;
+       std::string str = getString(buf, index);
+       os << "\"" << str << "\"";
+       os << " (aka [";
+       unsigned int size=str.size();
+       for(unsigned int i=0; i < size; i++) {
+           os << (int)str[i];
+           if(i < size-1)
```

Comment:
not a must, but since operations are optimized for comparing to zero and especially because there is no need to evaluate size and no need to subtract one per loop, this is cheaper expressed as (moved up a line):
```c
  if (i > 0) { os << ", "; }
```